### PR TITLE
Add accountID resolving logic for identity resolvers

### PR DESCRIPTION
**Pull request by wty-Bryant**
Add accountID resolving logic to allow different identity providers (e.g. ssocred/stscred/processcred/staticcred provider and env/shared config) extract aws account ID field from their retrieved credentials.

This feature will be cooperate with several remaining branches to support aws account ID endpoint2.0 routing.

### CODE REVIEW

File: .changelog/9166aec5123a472ab3d80afbcae6de9b.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "id": "9166aec5-123a-472a-b3d8-0afbcae6de9b",
+    "type": "feature",
+    "collapse": true,
+    "description": "Allow identity resolvers to resolve aws account id",
```
Comment (lucix-aws) [inline .changelog/9166aec5123a472ab3d80afbcae6de9b.json:5]: I think we can omit this entry in favor of a final one at the end.

File: aws/credentials.go [hunk 2]
```diff
@@ -90,6 +90,9 @@ type Credentials struct {
 	// The time the credentials will expire at. Should be ignored if CanExpire
 	// is false.
 	Expires time.Time
+
+	// AWS Account ID resolved from identity and used for optional endpoint2.0 routing
```
Comment (lucix-aws) [inline aws/credentials.go:94]: We don't need to project what it's used for here:

```suggestion
	// The ID of the account for the credentials.
```

File: config/env_config.go [hunk 3]
```diff
@@ -80,6 +80,8 @@ const (
 	awsRequestMinCompressionSizeBytes = "AWS_REQUEST_MIN_COMPRESSION_SIZE_BYTES"
 
 	awsS3DisableExpressSessionAuthEnv = "AWS_S3_DISABLE_EXPRESS_SESSION_AUTH"
+
+	awsAccountID = "AWS_ACCOUNT_ID"
```
Comment (lucix-aws) [inline config/env_config.go:84]: `awsAccountIDEnv`?

File: config/shared_config.go [hunk 4]
```diff
@@ -115,6 +115,8 @@ const (
 	requestMinCompressionSizeBytes = "request_min_compression_size_bytes"
 
 	s3DisableExpressSessionAuthKey = "s3_disable_express_session_auth"
+
+	accountID = "aws_account_id"
```
Comment (lucix-aws) [inline config/shared_config.go:119]: `accountIDKey`?

File: credentials/processcreds/provider.go [hunk 5]
```diff
@@ -167,6 +167,9 @@ type CredentialProcessResponse struct {
 
 	// The date on which the current credentials expire.
 	Expiration *time.Time
+
+	// The aws account ID for this op and could be used for endpoint2.0 routing
```
Comment (lucix-aws) [inline credentials/processcreds/provider.go:171]: Same from elsewhere, don't need to describe

**Review by isaiahvita — APPROVED** [review]

File: credentials/stscreds/web_identity_provider.go [hunk 6]
```diff
@@ -145,6 +152,14 @@ func (p *WebIdentityRoleProvider) Retrieve(ctx context.Context) (aws.Credentials
...
 	}
 	return value, nil
 }
+
+func getAccountID(assumedRoleUser *types.AssumedRoleUser) string {
+	if arn := assumedRoleUser.Arn; arn != nil && len(*arn) > 0 {
+		return strings.Split(*arn, ":")[4]
+	}
+	return ""
+}
```
Comment (lucix-aws) [inline credentials/stscreds/web_identity_provider.go:165]: You're still liable to panic on out-of-bounds this way, I'd restructure:

```suggestion
func getAccountID(u *types.AssumedRoleUser) string {
	if u.Arn == nil {
		return ""
	}

	parts := strings.Split(*u.Arn)
	if len(parts) < 5 {
		return ""
	}

	return parts[4]
}
```

Comment (wty-Bryant) [reply credentials/stscreds/web_identity_provider.go:165]: Okay I thought the arn format would be validated before the resolving. Will change that

**Review by lucix-aws — APPROVED** [review]

**Pull request merged**